plugins/autodim.koplugin/main.lua

```diff
@@ -0,0 +1,267 @@
+--[[--
+Plugin for automatic dimming of the frontlight after an idle period.
+
+@module koplugin.autodim
+--]]--
+
+local Device = require("device")
+local Event = require("ui/event")
+local FFIUtil = require("ffi/util")
+local SpinWidget = require("ui/widget/spinwidget")
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local TrapWidget = require("ui/widget/trapwidget")
+local time = require("ui/time")
+local util = require("util")
+local _ = require("gettext")
+local C_ = _.pgettext
+local T = FFIUtil.template
+
+local DEFAULT_AUTODIM_STARTTIME_M = 5
+local DEFAULT_AUTODIM_DURATION_S = 5
+local DEFAULT_AUTODIM_FRACTION = 20
+local AUTODIM_EVENT_FREQUENCY = 2 -- in Hz; Frequenzy for FrontlightChangedEvent on E-Ink devices
+
+local AutoDim = WidgetContainer:new{ name = "autodim" }
+
+function AutoDim:init()
+    self.autodim_starttime_m = G_reader_settings:readSetting("autodim_starttime_minutes", -1)
+    self.autodim_duration_s = G_reader_settings:readSetting("autodim_duration_seconds", DEFAULT_AUTODIM_DURATION_S)
+    self.autodim_fraction = G_reader_settings:readSetting("autodim_fraction", DEFAULT_AUTODIM_FRACTION)
+
+    self.last_action_time = UIManager:getElapsedTimeSinceBoot()
+
+    self.ui.menu:registerToMainMenu(self)
+    UIManager.event_hook:registerWidget("InputEvent", self)
+
+    self:_schedule_autodim_task()
+    self.isCurrentlyDimming = false -- true during or after the dimming ramp
+
+    self.trap_widget = TrapWidget:new{
+        dismiss_callback = function() self:restoreFrontlight() end,
+    }
```

Comment:
You could just create it when you're about to UIManager:show() it - instead of having a long-life one made in init.
It may word for this simple widget when you don't use more properties (like text), but there's usually no contract that our widgets can be show() then close() then show() again.
So, best to keep here the widget small lifecycle and not re-use it (as you don't really needed it). You can still store the last create instance as self.trap_widget to be able to close it in another method.